### PR TITLE
domain: close slow query channel after closing session pool

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

If slow query channel is closed before session pool, some session's goroutine may still
writing to the channel. Writing to a closed channel would cause TiDB panic.

image

### What is changed and how it works?

* Change the close order, close slow query channel after closing session pool.
* Drain the channel data before close the channel.
* Assume that after `domain.Close()` is called, there will be no more session executing.

### Check List 

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @crazycs520  @winkyao @coocood

File: domain/domain.go [hunk 1]
```diff
@@ -457,8 +457,8 @@ func (do *Domain) Close() {
 	if do.etcdClient != nil {
 		terror.Log(errors.Trace(do.etcdClient.Close()))
 	}
-	do.slowQuery.Close()
 	do.sysSessionPool.Close()
+	do.slowQuery.Close()
```
Comment (zimulala) [inline domain/domain.go]: Is it enough to put “do.slowQuery.Close()” after line462?

Comment (tiancaiamao) [reply domain/domain.go]: `do.slowQuery.Close()` will make goroutine exit and WaitGroup count -1.
If this line is moved to line 462,`do.wg.Wait()` would wait forever.

Comment (zimulala) [reply domain/domain.go]: Ignore it.

Comment (zimulala) [reply domain/domain.go]: But I think it's useless. After this function is executed, "updateStatsWorker" may still be called by another goroutine. This problem still exists.
Maybe we can replace "ctx"(the updateStatsWorker's argument) with the session created in the sysSessionPool.

File: domain/topn_slow_query.go [hunk 2]
```diff
@@ -196,6 +196,14 @@ func (q *topNSlowQueries) QueryTop(count int, kind ast.ShowSlowKind) []*SlowQuer
 }
 
 func (q *topNSlowQueries) Close() {
+	done := false
+	for !done {
+		select {
+		case <-q.ch:
```
Comment (winkyao) [inline domain/topn_slow_query.go]: Why do we need to drain the channel?

File: domain/domain.go (hunk 1 above)
Comment (tiancaiamao) [reply domain/domain.go]: Do you mean after session pool is closed, `updateStatsWorker`  will be still executing ? @zimulala

Comment (crazycs520) [reply domain/domain.go]: Ye. Because the `session` used in statistic bootstrap is not get from `session pool`. @tiancaiamao

Comment (tiancaiamao) [reply domain/domain.go]: Current session pool implementation will **block waiting resource to be put back**, so after the pool is closed, there won't be system session running. @crazycs520 @zimulala

File: domain/topn_slow_query.go (hunk 2 above)
Comment (tiancaiamao) [reply domain/topn_slow_query.go]: Removed.

File: domain/topn_slow_query.go [hunk 3]
```diff
@@ -115,6 +115,11 @@ type topNSlowQueries struct {
 	period   time.Duration
 	ch       chan *SlowQueryInfo
 	msgCh    chan *showSlowMessage
+
+	mu struct {
+		sync.RWMutex
+		closed bool
```
Comment (crazycs520) [inline domain/topn_slow_query.go]: how about use `atomic` instead of `mutex`.

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: atomic can't protect multiple operations. @crazycs520 

```
modify atomic flag
close(ch)

check atomic flag
<-ch
```

Take this order for example:

1. check atomic flag = success
2. modify atomic flag = closed
3. close(ch)
4. <-ch  panic!

**Pull request merged**